satellite/satellitedb/repairqueue.go

```diff
@@ -31,38 +34,56 @@ func (r *repairQueue) Enqueue(ctx context.Context, seg *pb.InjuredSegment) error
 	return err
 }
 
-func (r *repairQueue) Dequeue(ctx context.Context) (pb.InjuredSegment, error) {
-	// TODO: fix out of order issue
-	tx, err := r.db.Open(ctx)
+func (r *repairQueue) Dequeue(ctx context.Context) (seg pb.InjuredSegment, err error) {
+	// note: BeginTx(ctx, &sql.TxOptions{Isolation: ...) didn't work
+	// so we're using SQL 'FOR UPDATE' below instead
+	tx, err := r.db.DB.Begin()
 	if err != nil {
-		return pb.InjuredSegment{}, Error.Wrap(err)
+		return seg, Error.Wrap(err)
 	}
-
-	res, err := tx.First_Injuredsegment(ctx)
+	defer func() {
+		if err == nil {
+			err = Error.Wrap(tx.Commit())
+		} else {
+			err = Error.Wrap(utils.CombineErrors(tx.Rollback(), err))
+		}
+	}()
+	//get top
+	selectSQL := ""
+	switch t := r.db.DB.Driver().(type) {
+	case *sqlite3.SQLiteDriver:
+		selectSQL = `SELECT id, info FROM injuredsegments LIMIT 1`
+	case *pq.Driver:
+		selectSQL = `SELECT id, info FROM injuredsegments LIMIT 1 FOR UPDATE`
+	default:
+		return seg, fmt.Errorf("Unsupported driver %t", t)
+	}
+	rows, err := tx.Query(selectSQL)
 	if err != nil {
-		return pb.InjuredSegment{}, Error.Wrap(utils.CombineErrors(err, tx.Rollback()))
-	} else if res == nil {
-		return pb.InjuredSegment{}, Error.Wrap(utils.CombineErrors(storage.ErrEmptyQueue.New(""), tx.Rollback()))
+		return seg, err
 	}
-
-	deleted, err := tx.Delete_Injuredsegment_By_Id(
-		ctx,
-		dbx.Injuredsegment_Id(res.Id),
-	)
+	//defer func() { err = errs.Combine(err, rows.Close()) }()
+	if !rows.Next() {
+		rows.Close()
+		return seg, rows.Err()
+	}
+	var id int64
+	err = rows.Scan(&id, &seg)
 	if err != nil {
-		return pb.InjuredSegment{}, Error.Wrap(utils.CombineErrors(err, tx.Rollback()))
-	} else if !deleted {
-		return pb.InjuredSegment{}, Error.Wrap(utils.CombineErrors(Error.New("Injured segment not deleted"), tx.Rollback()))
+		rows.Close()
+		return seg, err
 	}
-	if err := tx.Commit(); err != nil {
-		return pb.InjuredSegment{}, Error.Wrap(err)
+	rows.Close()
+	//delete
+	res, err := tx.Exec(r.db.Rebind(`DELETE FROM injuredsegments WHERE id = ?`), id)
```

Comment:
In the PostgreSQL case, we can actually do this whole function in one simpler query:

```go
func (r *repairQueue) Dequeue(ctx context.Context) (seg pb.InjuredSegment, err error) {
    err = r.db.DB.QueryRowContext(ctx, `
        DELETE FROM injuredsegments
            WHERE id = (
                SELECT id FROM injuredsegments
                    ORDER BY something
                    FOR UPDATE SKIP LOCKED
                    LIMIT 1
            )
            RETURNING info
    `).Scan(&seg)
    return seg, err // err will be sql.ErrNoRows when there were no rows left in the table
}
```

This has advantages besides being shorter: no explicit transaction is necessary, and the query optimizer will be able to see what you're doing and do it more efficiently!

Buuuuuuuut for the sqlite case, I'm pretty sure you won't be able to do `DELETE ... RETURNING`. So you'd still need to have a transaction, do a QueryRowContext to get both the id and the info, and do a delete statement and check the errors and RowsAffected. So maybe I haven't saved you any code. But at least it would work better in PG land...